prov/rxm/src/rxm_cq.c

```diff
@@ -247,8 +247,12 @@ static inline int rxm_finish_sar_send(struct rxm_tx_entry *tx_entry)
 	tx_buf->pkt.ctrl_hdr.type = ofi_ctrl_data;
 	rxm_tx_buf_release(tx_entry->ep, tx_buf);
 
-	if (!--tx_entry->seg_cnt)
-		return rxm_finish_send_nobuf(tx_entry);
+	if (!--tx_entry->seg_cnt) {
+		RXM_LOG_STATE(FI_LOG_CQ, tx_buf->pkt,
+			      RXM_SAR_TX, RXM_SAR_ACK_WAIT);
+		tx_entry->state = RXM_SAR_ACK_WAIT;
```

Comment:
We can add a sequence number to the msg_id to prevent aliasing of different messages under the same msg_id.  That could help us avoid implementing a acknowledgement mechanism. What do you think?

Also, I don't quite understand why we get this issue in the first place. Messages are expected to arrive in order at the receiver. So if two medium sized messages have same msg_id their segments should complete in order.

Reply:
Yes, you're right, the finalization protocol isn't needed here at all.

I've removed finalization protocol as a not needed anymore.
So, I found a bug in my previous code. The finalization just hides this bug.